### DISCUSSION TITLE
Manage one database per git branch

### DISCUSSION THREAD

**Discussion by angererc** [theory/sqitch#626] [Q&A] [answered]
Hi,

We would like to find a friction-less way to manage a separate database per git branch. 

Eg imagine the master branch, accompanied by a master DB and sqitch config and plan to set up the master DB. Now I start working on a feature X and start a new GIT feature branch X. I don't want to get in the way of my fellow developers, therefore I also create a new DB X. When I'm done implementing X, I want to merge back into master and delete the X DB. 

Now how do I best reconfigure sqitch in the X branch to use the new database X? If I add a target, then merging the changes back into master will have this left-over X target of the (now defunct) X branch, right? If I change the original target to point to the X DB, then merging into master will override the master target. I somehow want to temporarily, during the lifetime of a branch, override a target's database URI or (alternatively) create a temporary target that doesn't get merged back. (I hope this makes semse:) Any thoughts?

**Comment by theory** [comment, marked as answer]
Most engines support [environment variables](http://sqitch.org/docs/manual/sqitch-environment/) and defaults that specify database connection information. Postgres, for example, will fall back on your local username as the database name and username. So if you set a target like this:

``` sh
sqitch add target local-dev db:pg:localhost/
```

Then for a user named `bob` it would connect to `db:pg:localhost/bob` and for a user named `sue` it would connect to `db:pg:localhost/sue`. All you have to do is remember to use it each time

``` sh
sqitch deploy local-dev
```

Or set `$SQITCH_TARGET` to that target:

``` sh
export SQITCH_TARGET=local-dev
sqitch deploy # uses local-dev
```

Of course, if you're setting environment variables, you can just do that generally, esp if youre engine doesn't natively support the flexibility you require. For example, you could use a tool like [direnv](https://direnv.net) and commit an `.envrc` file to the root of your project with something like this:

```
export SQITCH_TARGET=db:pg://localhost/myproj_$(git branch --show-current)
```

Then every time you check out a new branch, as long as you've told direnv to trust the file, it will automatically reset the environment variable with the branch name as part of the database name.

HTH!

**Reply by angererc** [reply]
Thank you for your answer. I think your direnv suggestion will be workable for us, we'll test this out. I still think it would be nice to have such a feature built into sqitch (git-branch-aware DB targets) so one can tie the dev database to the git feature branch you are working on. But I also understand that that's probably not easy to do in a VCS independent way, therefore the env variable workaround seems like a good solution.

**Reply by theory** [reply]
That's a fair request. I've long wanted better integration with VCS software, but aside from [sqitch checkout](http://sqitch.org/docs/manual/sqitch-checkout/) for Git only, I've not had the chance and there hasn't been much demand TBH. I could perhaps see some additional branch awareness that, say, acted as an environment variable for the branch name to use in target or some such, but it would require a fair bit of thinking through, since branches can be local or remote or in different remotes, but even some sort of simply support for environment variable interpolation in [Config::GitLike](/bestpractical/config-gitlike/) would go a long way, especially for CI/CD pipelines that tend to set an environment variable to a branch name.
